rfcs/007 - Code review infra changes.md

```diff
@@ -0,0 +1,60 @@
+# 007 - Code review infrastructure changes RFC
+
+### Proposed
+
+2022-12-16
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- David Ellis <david@iasql.com>
+- Yolanda Robla <yolanda@iasql.com>
+- Alejandro Guillen <alejandro@iasql.com>
+- Mohammad Pabandi <mohammad@iasql.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Luis Fernando De Pombo <l@iasql.com>
+
+## Summary
+
+Eng teams move over from using cloud UIs to IaC because cloud infrastructure changes can lead to outages more often than business logic changes. As a result, it is considered a DevOps best practice to code review infrastructure changes using IaC tools and show the `preview` of the resulting change in pull requests to raise the quality bar and reduce the possibility of mistakes. Cloud infrastructure can be code reviewed with IaSQL using any migration system that supports non-schema, or data, migrations (some ORMs like Prisma only support schema changes) by wrapping each migration in an IaSQL transaction. However, through our research and dogfood efforts of IaSQL we have not found a migration system that fulfills the requirements needed to code review changes with IaSQL:
+- Allow data-only, `up` migrations
+- Allow plain SQL or have an introspective ORM
+- Free and ideally open source
+- Baseline/ignore certain migrations
+
+Data migrations are needed for IaSQL, more so than the traditional schema migrations, and require baselining migrations to avoid duplicate record violations when IaSQL databases are recreated from scratch (e.g. connected and disconnected to the same cloud account) or when changes happen outside of IaSQL and are “synced” from the cloud by the cron job. Flyway comes closest to meeting all our requirements. However, it is not possible to baseline/ignore existing migrations on their free product.
+
+## Proposal
+
+The proposal is to implement a super lightweight, open-source plain SQL migration system in its own public repository, such that it could be used separately without IaSQL. The feature set for the initial MVP version:
+- A downloadable CLI with a single `migrate` command that works with Linux, so it is compatible with CI/CD
+- Support PostgreSQL 14
+- Support `up` versioned migrations with the following format `migrations/V0__********.sql`, `migrations/V1__********.sql`
+- Be able to baseline migrations from a given version so running `migrate` ignores migrations before a given version
+- Three config values are provided to `migrate` through environment variables: baseline version (defaults to 0), migration folder location (defaults to `migrations/` in the current directory), and a required PostgreSQL connection string
+
+## Alternatives Considered
```

Comment:
I am starting to get an idea for an alternative in my mind that may supplement or supplant the main proposal. Stay tuned for a big comment in a couple of hours :)

Reply:
So one of the advantages of IaSQL is that when you write statements `INSERT`, `UPDATE`, or `DELETE` something, the database can error back at you telling you you're missing some required column, or that you're trying to set a column to an invalid value, or you're trying to remove something that still has something dependent on it.

It's a level even above what Pulumi is capable of in typed languages because the "type safety" can extend to particular records/values actually in your cloud account that just static typing would be unable to provide.

This "super power" is completely lost in a traditional migration system because the file is static and unrelated to your database. You write SQL into a migration file and you can only confirm that it will actually work by running it (against a test database or the real thing).

It does give you the ability to let others review the change and potentially sign off on it or not, but beyond their own ability to spot typos, it is unlikely they can catch any errors in what you wrote, beyond "does this seem to implement what we're trying to accomplish in our infrastructure."

But if we put the database in the loop, particular IaSQL, we could make something even better.

Using what we're thinking of in RFC 006, where we can take a set of audit log records and turn them into SQL, users could run `SELECT iasql_begin();`, then write their proposed changes in the dashboard or their preferred SQL editor of choice, and could then use the proposed `iasql_get_sql_since()` or perhaps some other function that automatically grabs the current IaSQL transaction's changes and get the SQL statements to produce the changes they are proposing, with any mistakes they made along the way automatically filtered out because those statements would not have mutated any of the tables, finally the user can `iasql_preview()` to see that the changes do what's expected of them and `iasql_rollback()` to revert the changes and drop the transaction. The generated SQL can be copy-pasted into their migration system of choice for external review, but with much more confidence that it would actually work.

But we can do better than that! Since we are already in the IaSQL database for authoring a change that should be reviewed, there's no reason why we can't have helper functions to handle all of this.

`iasql_review_changes()` would grab the audit log records starting from the current transaction, generate the SQL, generate the execution preview, and save all of that into a new table (that would only need three columns `id: int, sql: text, expected_operations: json`), then it would `iasql_revert()` to before the transaction started and return the ID as the output.

Others can call `iasql_view_changes(id)` to get the change in question and see it displayed in their own SQL connection, and then you can either `iasql_reject_changes(id)` (which would delete the row from the table) or `iasql_approve_changes(id)` (which would start a new transaction, run the SQL, then `iasql_commit()` it).

We could potentially make a fifth function that bakes in Github support `iasql_pr_changes(id, gh_repo, gh_pat)` to auto-generate a github PR with the description being a markdown version of the `iasql_view_changes` output, and merging the PR *somehow* securely calls `iasql_approve_changes(id)`, but not sure how much extra work that would be, and there's a bunch of weirdness around that PR getting out-of-sync with the database if someone calls `iasql_reject_changes(id)` directly in the database and someone else later merges the PR, so I'm less into that one as written (but perhaps it could be refined into something better, and maybe even inclusive to Gitlab and others?)

In any case, the original 4 proposed functions would just be 4 pure SQL functions with 1 table, and could be in a new `iasql_review` module that users can optionally install if they want that sort of thing, which would only depend on the other `iasql_*` modules.

Reply:
I also had this idea of "running migrations through IaSQL and not pure Postgres" since if we go with Postgres, there will be some complexities. And Luisfer had proposed the "baselining" requirement since if we don't baseline migrations, we might run into errors on `INSERT`s (if the object already exists - for example) which is one of the complexities.
So the idea of "running migrations through IaSQL helper functions" comes into the mind. I was thinking of a way to *somehow* know if an `INSERT` should be executed or it has already been executed. That'd be possible because we keep track of changes we push/pull to/from the cloud and we might be able to see if a change has already been pushed to the cloud. Which opens our hand to appropriately handle the situation (which if the rows are directly `INSERT`ed to the Postgres, isn't possible).

But the problem in my mind was this: having a function like `run_iasql_migration('INSERT INTO bucket () VALUES ()')` is an ugly method and providing an interface that treats SQL queries like strings is a bad idea.

David's proposed idea is a good one and eliminates the ugliness I mentioned above. Also we can programmatically handle approvals for a set of changes (and I suggest we also add multiple users for the same database - e.g. `david` and `mohammad` as two engineers maintaining an infrastructure). And changes can be approved if at least a number of users have approved them.

I don't see any downside for "running migrations through IaSQL helper functions" and not directly.